### PR TITLE
support for nuttx RTOS

### PR DESCRIPTION
**Pull request by jafrado**
Please find attached my working port to Nutt-X on STM32. I noticed that bmp3_defs.h and bmp.h could be merged. Also, the dev structure should be pass to the read routines directly so that they may be accessed through an opaque pointer (the "dev" structure on the target OS). Thank you very much for the great development and clean coding.

### CODE REVIEW

File: bmp3.c [hunk 1]
```diff
@@ -1323,9 +1339,12 @@ int8_t bmp3_get_sensor_data(uint8_t sensor_comp, struct bmp3_data *comp_data, st
 
     /* Array to store the pressure and temperature data read from
      * the sensor */
-    uint8_t reg_data[BMP3_P_T_DATA_LEN] = { 0 };
-    struct bmp3_uncomp_data uncomp_data = { 0 };
+    uint8_t reg_data[BMP3_P_T_DATA_LEN];
+    struct bmp3_uncomp_data uncomp_data;
```
Comment (jafrado) [inline bmp3.c:1331]: Not supported on all compilers, memset instead.

File: bmp3.c [hunk 2]
```diff
@@ -2649,3 +2668,237 @@ static int8_t get_err_status(struct bmp3_dev *dev)
 
     return rslt;
 }
+
```
Comment (jafrado) [inline bmp3.c:2658]: The below may be removed; and you may keep it as a sample if you wish.

File: bmp3.h [hunk 3]
```diff
@@ -46,13 +46,777 @@
...
 /*! CPP guard */
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+#include <stdint.h>
+#include <stddef.h>
+
+/*************************** Common macros   *****************************/
+
```
Comment (jafrado) [inline bmp3.h]: Below references Linux kernel header files, I adapted it to a non-Linux kernel environment.

File: bmp3.h [hunk 4]
```diff
@@ -46,13 +46,777 @@
...
+    struct bmp3_settings settings;
+
+    /*! Sensor and interrupt status flags */
+    struct bmp3_status status;
+
+    /*! FIFO data and settings structure */
+    struct bmp3_fifo *fifo;
+
+    FAR struct bmp3_dev *flink;          /* Supports a singly linked list of drivers */
+    FAR struct i2c_master_s *i2c;        /* I2C interface */
```
Comment (jafrado) [inline bmp3.h]: The i2c_master_s may be removed, it's specific to Nutt-X

File: bmp3.h [hunk 5]
```diff
@@ -317,6 +1081,21 @@ int8_t bmp3_get_status(struct bmp3_dev *dev);
  */
 int8_t bmp3_set_fifo_watermark(const struct bmp3_dev *dev);
 
+/*!
```
Comment (jafrado) [inline bmp3.h:1081]: Nut-X register interface below.

**Review by jafrado — COMMENTED** [review]
notes for integration.

**Comment by BST-Github-Admin** [comment]
Hi @jafrado, 
We would like to avoid platform specific code in the driver which will prevent it from being compiled for other targets. Nevertheless, there are some comments that you have made that are still valid. In order to validate them and include them in our internal design guidelines, can you comment on which compiler you use and the related build options?

**Comment by jafrado** [comment]
Sure, I can rip out the nuttx defines. Would you like me to resubmit a patch without that?

Using GCC from ARM arm-none-eabi-gcc -v reports:

gcc version 6.3.1 20170620 (release) [ARM/embedded-6-branch revision 249437] (GNU Tools for ARM Embedded Processors 6-2017-q2-update) 

Compiler options:
```
arm-none-eabi-gcc -c -fno-builtin -Wall -Wstrict-prototypes -Wshadow -Wundef -g -mcpu=cortex-m7 -mthumb -mfloat-abi=soft bmp3.c
```
You can download the stable (old) Nutt-X compiler here:

Download link [GCC ARMReleased: June 28, 2017](https://developer.arm.com/-/media/Files/downloads/gnu-rm/6-2017q2/gcc-arm-none-eabi-6-2017-q2-update-linux.tar.bz2?revision=2cc92fb5-3e0e-402d-9197-bdfc8224d8a5?product=GNU%20Arm%20Embedded%20Toolchain,64-bit,,Linux,6-2017-q2-update)

**Comment by jafrado** [comment]
I just pushed another set of mods, removed the system dependencies and added a Makefile for you to test build.

**Comment by jafrado** [comment]
The array initializers below, they are not supported on older compilers:

uint8_t reg_data[BMP3_P_T_DATA_LEN] = { 0 };
struct bmp3_uncomp_data uncomp_data = { 0 };

**Comment by BST-Github-Admin** [comment]
Hi @jafrado ,

can you change to a newer dialect by adding the option `-std=c99`. I hope that helps.

**Comment by jafrado** [comment]
Hi @BoschSensortec, thank you for your reply. NuttX follows C89 and all code outside of architecture specific directories must be compatible with C89. So adding that to fix an initialization statement written in C99 won't work because there are a myriad of spots which simply won't compile. 

Related,  the macros 
```
S{8,16,32,64}_C,  U{8,16,32,64}_C 
```
are kernel type macros found in:
```
linux/include/asm-xxx/int-ll64.h 
```
and should also be defined under the 
```
#ifdef __KERNEL__ 
```
which means this driver has only every been built for Linux!

Removing this bmp_defs.h gets rid of additional code needed for the kernel case, and I have
provide options which will work on any embedded ARM without an MMU.
Please have a look.

**Comment by BST-Github-Admin** [comment]
Updatef SensorAPI architecture.

**Pull request closed without merging**